platforms/core-runtime/INSTRUMENTATION.md

```diff
@@ -0,0 +1,437 @@
+# Instrumentation and bytecode upgrades infrastructure
+
+
+## Abstract
+
+This README describes the current state instrumentation and bytecode infrastructure used for configuration cache instrumentation and byte code upgrades used for property migration.
+
+
+## Background
+
+With the introduction of configuration cache we got a requirement to track user inputs for the configuration phase. 
+Inputs could be environment variables, files, system properties and so on. 
+These inputs are often read via an API that Gradle doesn’t control, e.g. Java API via System.getenv().
+
+So to track these inputs we had to find a way to track these usages. 
+And that is why we started instrumenting bytecode.
+
+With a requirement to migrate Gradle core APIs to lazy properties, we were looking at how we could make old plugins compatible with new Gradle core APIs. 
+We decided that upgrading old compiled bytecode would work well. And reusing configuration cache bytecode instrumentation seemed like a good tool to reuse.
+
+So configuration cache instrumentation and API upgrades are using the same infrastructure now.
+
+
+## Functional design
+
+Imagine we have in Gradle core a task:
+```
```

Comment:
:x: Let's add the languages here so that we get syntax highlighting.

Reply:
Done